### DISCUSSION TITLE
Using Embedded Python with BWAPI

### DISCUSSION THREAD

**Discussion by Sneakyabc** [bwapi/bwapi#906] [Q&A]
Hi there!

I've been searching for an appropriate place to ask this question, so hopefully this is a good place and someone can generously help me. I've been trying to develop an AI using pytorch to use in a BWAPI module bot. I've created a simple simulation environment for SC:BW in C++ so that I could train the pytorch models in my simulation rather than having to fiddle with constantly going in and out of the actual game. To use pytorch I've been running python code directly from C++ (#including <Python.h>), this link shows an example of how I've been using this:

https://stackoverflow.com/questions/37399930/how-do-you-run-a-python-script-from-a-c-program

I've finally gotten the AI model to a place where I'm ready to try using it within BWAPI, but I've run into some complications. I can include <Python.h> in the Example AI Module project and get it to run in the game no problem. But the moment I put in any function that actually makes use of Python.h (for instance, Py_Initialize() or PyObject_CallMethodObjArgs()) the AI module fails to load when starting the game ("Failed to load the AI Module ExampleAIModule.dll").

As this wasn't working, I also tried using TorchScript, which is a C++ library for importing wrapped Torch models, but I couldn't get that to compile.

Would any geniuses out there happen to know if this sort of thing should work with BWAPI, and what I might be missing? I'm not too familiar with compiling projects into DLLs so I'm hoping that it's a relatively simple fix but I know how complicated and messy these libraries and environments can be. I can provide more information if needed, as a quick explanation: I'm on windows 10, using VS 2017, I have 'Python310-32\include' folder included in the C/C++ Additional Include Directories options for the ExampleAIModule project and 'Python310-32\libs' added to the Linker Additional Library Directories options for the same project. With this and #include <Python.h> it compiles and runs in game just fine. But if I uncomment "Py_Initialize();" in the Module it will compile just fine and then fail to load once in game.

I'm guessing I'm not the first to try this sort of approach. I love programming and I love Starcraft so I'd be very bummed to abandon this project now, huge thanks for any help in advance.

![bwapi_1](https://github.com/bwapi/bwapi/assets/39353207/cf8e0fb8-0f8f-4168-ab22-1c27d3c742f1)
![bwapi_2](https://github.com/bwapi/bwapi/assets/39353207/d77879d7-3021-498a-a5f6-ad30bc95bece)
![bwapi_3](https://github.com/bwapi/bwapi/assets/39353207/7ea5dcad-c80b-452e-995e-13769409959c)
![bwapi_4](https://github.com/bwapi/bwapi/assets/39353207/80d70e86-c80d-4804-be9c-1c8b4cac4258)

**Comment by N00byEdge** [comment]
I would recommend opening your bot .dll in `depends.exe` to check if all its dll dependencies (like torch) are satisfied. You have to have all dlls your bot depends on in the DLL include path, for example in the same AI directory.

**Reply by Sneakyabc** [reply]
Hmm okay, by `depends.exe` I'm assuming you're talking about this Dependency Walker program I found online. I'm not familiar with this at all so I'm assuming this should show me what included .dll files a .exe or .dll file has packaged inside of it? I see a bunch of system32 stuff here and I assume I should also be seeing the Python .dll's as well if I export it correctly (I think those API-MS-* warnings are false positives from reading online). 

So if I'm understanding this, I need to double check my settings with VS 2017 that the required .dll files are getting packaged into the final .dll that is exported? If they don't need to get packaged into the final `ExampleAIModule.dll` then I don't see what the problem would be since the project compiles just fine. If I'm right about that, I'm not sure how I would tell VS 2017 to export that way. If I'm wrong I'd love to be corrected.

![bwapi_5](https://github.com/bwapi/bwapi/assets/39353207/6c100ed8-1aaf-4c52-b7aa-e7e7a5655f8e)

I looked through the settings of the projects/solution a bit and found this option that seemed interesting to me. In the properties menu for the solution (not project), under Project Dependencies, ExampleAIModule wasn't checked. I checked it off, rebuilt and opened the new .dll with `depends` and now I see `PYTHON310_D.DLL` that I didn't see before. The bot still didn't load, but I'm guessing I'm on the right path, so I'll keep fiddling with things to see if I can figure it out and I'll keep an eye on here for advice. Thanks again.

![bwapi_6](https://github.com/bwapi/bwapi/assets/39353207/19fd7281-5366-4912-896e-cf8b0343637f)
![bwapi_7](https://github.com/bwapi/bwapi/assets/39353207/6990f87d-e9bf-40e3-8208-6937975c2a93)

**Reply by N00byEdge** [reply]
Yeah, you probably need to bundle `PYTHON310_D.dll` with your bot if you want to link against a debug python dynamically here. Yes the ones on the first file are false positives. The python one is not.

**Reply by Sneakyabc** [reply]
Inside the `libs` and `include` folders for Python I don't even see any DLL files, just headers and `.libs`. I'm guessing either I'm missing the DLLs or the DLLs would be derived from the `.libs`? I don't know the nitty-gritty details of how all of these libraries and DLLs really work when compiling. I'll just have to do some googling. If I can figure this out I'll post the solution.

**Reply by N00byEdge** [reply]
The python debug dll would either be part of a debug build that you download or, if you are building python from source, created when you built python. If you used a release build of python, you have to build your bot in release mode and link against the release python library.

**Reply by Sneakyabc** [reply]
Okay... I'm not sure if this is progress but I've got things to report. Looking around randomly I noticed that the `python310_d.dll` Dependency Walker is talking about is located one folder up from the `\python310-32\libs` folder. I tried copying the DLL into the libs folder and nothing changed. However, after building, if I put the `python310_d.dll` in the same folder as the `ExampleAIModule.dll`, and then open the AI DLL with Dependency Walker, it DOES see the DLL and its functions.

___
![bwapi_9](https://github.com/bwapi/bwapi/assets/39353207/82c7a747-bee1-4159-b4a3-1d4ebf36d378)
___

Even if I drop both DLL files into the `bwapi-data\AI` folder the bot still doesn't load, though. Perhaps the BWAPI program is looking elsewhere for it?

From my understanding reading online, a DLL is dynamic, meaning the necessary function definitions are loaded at runtime. So the necessary DLL file must be present on a system in order for a program that needs its functions to run. Whereas a LIB file is (or at least can be) static, meaning the function definitions are pulled at compile time, so you don't need a DLL to run. Since I am pointing VS to LIB files and not the DLLs in the project settings, shouldn't it be compiling in a manner where I don't need the DLL Dependency Walker is seeing? Once I've built the bot, shouldn't I only have/need the `ExampleAIModule.dll` that is output? If I am expected to provide both DLL files, where would BWAPI be expecting to find it? 

I feel like this is some kind of problem with how VS is linking everything together that I'm just not fully grasping.

**Reply by Sneakyabc** [reply]
Whelp, trying random things I figured out that if I put the `python310_d.dll` file in the `Program Files (x86)\StarCraft` folder, the AI module runs. I have VS attached to StarCraft here, which I couldn't do before because the module wasn't loading.

___
![bwapi_10](https://github.com/bwapi/bwapi/assets/39353207/964d2865-c380-4e3e-8e12-6963c1fb0dea)
___

So I suppose this works, but why do I need the DLL in the StarCraft directory? I'm confused why the project isn't building the necessary functions into the AI Module's DLL so that I don't need another DLL for it to run. Is it common for other BWAPI bots to submit with additional DLLs they require? I'm guessing there's project settings I need to mess with to get that output.

I feel a little bad this didn't seem to be an issue with BWAPI at all and was just issues with me getting the project to link in the way I expect but hopefully if someone else has an issue like this this post will help. And thanks for the help, I know it's time consuming trying to follow what someone else is experiencing.

**Reply by N00byEdge** [reply]
> I'm confused why the project isn't building the necessary functions into the AI Module's DLL so that I don't need another DLL for it to run

That's called static linking and requires a different build configuration. You should check the python resources to see how to statically link against python.

**Reply by Sneakyabc** [reply]
Yeah I have a few tabs queued up to read up on forcing static linkage. Rethinking how this should all work I think I don't want to use the Python libraries directly because then if I wanted to run my bot on another machine I'd have to ensure that other machine had python, numpy, pytorch, and everything installed on it. I think trying to get the TorchScript C++ library running would be better, but that is giving me unresolved reference errors and won't compile.
Always something to fix! Thanks for the help, I'll probably close the discussion since I think I understand why I was having problems before now.

**Comment by Sneakyabc** [comment]
Closed

**Reply by dgant** [reply]
If you get this project setup working I think folks would love to see it!
